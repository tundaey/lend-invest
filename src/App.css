.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

.Investments-title {
  margin-top: 0.5em
}

.invested-row {
  background-color: #f9f9f9
}
.invested-badge{
  background-color: green;
  display: inline-block;
  min-width: 10px;
  padding: 3px 7px;
  font-size: 12px;
  font-weight: 700;
  line-height: 1;
  color: #fff;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  border-radius: 10px;
}
