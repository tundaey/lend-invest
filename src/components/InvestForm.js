import React, {Component} from 'react';
import {Modal, Button, FormGroup, ControlLabel, FormControl} from 'react-bootstrap'

class InvestForm extends Component {
    constructor(){
        super()
        this.state = {
            investment: 0,
            loan: {}
        }
    }

    render(){
        return (
            <Modal show={this.props.show}>
                <Modal.Header>
                    <Modal.Title>Invest</Modal.Title>
                </Modal.Header>

                <Modal.Body>
                    <FormGroup controlId="formBasicText">
                        <ControlLabel>Enter Desired Investment(Â£)</ControlLabel>
                        <FormControl
                            type="number"
                            value={this.state.investment}
                            placeholder="0"
                            onChange={this.updateForm}
                        />
                    </FormGroup>
                </Modal.Body>

                <Modal.Footer>
                    <Button id="cancelButton" onClick={(event)=> this.props.closeModal(event)}>Cancel</Button>
                    <Button id="investButton" onClick={(event)=> this.updateLoan(event)}  bsStyle="primary">Invest</Button>
                </Modal.Footer>
            </Modal>
        )
    }
}

export default InvestForm