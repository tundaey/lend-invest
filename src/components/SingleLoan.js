import React from 'react';
import {Badge} from 'react-bootstrap'

const SingleLoan = (props) => (
    <tr className={props.loan.invested? 'invested-row ': ''}
        onClick={(event)=> props.onClick(event, props.loan)}>
        <td>{props.loan.id}</td>
        <td>{props.loan.title}</td>
        <td>{props.loan.tranche}</td>
        <td>{props.loan.available}</td>
        <td>{props.loan.annualised_return}</td>
        <td>{props.loan.term_remaining}</td>
        <td>{props.loan.ltv}</td>
        <td>{props.loan.amount}</td>
        <td><Badge bsClass={props.loan.invested? 'invested-badge': 'badge'}>{props.loan.invested? 'Invested': 'Eligible'}</Badge></td>
    </tr>
)

export default SingleLoan