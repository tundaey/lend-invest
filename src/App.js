import React, { Component } from 'react';
import './App.css';
import Loans from './components/Loans'
import {loans} from './current-loans.json'
import NavBar from './components/Navbar'
import InvestForm from './components/InvestForm'

class App extends Component {
  constructor(){
    super()

    this.state = {
      title: 'LendInvest',
      isModalOpen: false,
      loans: [],
      selectedLoan: {},
      possibleInvestments: []
    }
  }

  componentDidMount(){
    this.setState({loans: loans, possibleInvestments: loans})
  }

  render() {
    return (
      <div>
        <NavBar title={this.state.title}/>
        <div className="container">
          <div className="row">
            <div className="col-md-8">
              <br/>
              <div className="col-md-6"><h1>All Current Loans</h1></div>
              <div className="col-md-6 Investments-title">
                <h3>Possible Investments: <span>{this.state.possibleInvestments.length}</span></h3>
                </div>
            </div>
            <div className="col-md-11">
              <Loans 
                openModal={this.openModal}
                isModalOpen={this.state.isModalOpen} 
                closeModal={this.closeModal} 
                loans={this.state.loans}/>

              {this.state.isModalOpen 
                ? <InvestForm 
                closeModal={this.closeModal}
                processInvestment={this.processInvestment}
                selectedLoan={this.state.selectedLoan} 
                show={this.state.isModalOpen}/>
              : null}
            </div>
          </div>
        </div>
      </div>
    );
  }

  closeModal = (event)=>{
    event.preventDefault()
    this.setState({isModalOpen: false, selectedLoan: {}})
  }

  openModal = (event, loan)=>{
    event.preventDefault()
    this.setState({isModalOpen: true, selectedLoan: loan})
  }

  processInvestment = (loan, investment)=>{
    const available = parseFloat(loan.available.replace(/,/g, ''));
    const newAvailableAmount  = available - investment
    loan['available']= newAvailableAmount.toLocaleString('en')
    const newLoans = this.state.loans.map((l)=> {
      if(l.id === loan.id){
        loan['invested'] = true
        return loan
      }
      return l
    })
    
    this.updatePossibleInvestments(loan)
    this.setState({isModalOpen: false, loans: newLoans})
  }

  updatePossibleInvestments = (loan)=>{
    const investments = this.state.possibleInvestments
    const invested = investments.indexOf(loan)
    if(invested > -1){
      investments.splice(invested, 1)
      this.setState({possibleInvestments: investments})
    }
  }


}

export default App;
