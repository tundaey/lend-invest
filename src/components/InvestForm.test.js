import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import {loans} from '../current-loans.json'

import InvestForm from './InvestForm'
import SingleLoan from './SingleLoan'
import {Modal, Button, FormControl, Badge} from 'react-bootstrap'
import App from '../App'



test('renders the invest form in the modal properly', ()=> {
    const show = true
    const wrapper = shallow(<InvestForm show={show} />)
    const input = wrapper.find(FormControl)
    const buttons = wrapper.find(Button)
    const modal = wrapper.find(Modal)
    expect(modal.length).toBe(1)
    expect(input.length).toBe(1)
    expect(buttons.length).toBe(2)

})

test('User can close popup when the cancel button is called', ()=> {
    const AppComponent = mount(<App/>)
    const spy = jest.spyOn(AppComponent.instance(), 'closeModal');

    const show = true
    const wrapper = mount(<InvestForm closeModal={spy} show={show} />)
    const cancelButton = wrapper.find('#cancelButton').first()

    cancelButton.simulate('click')
    expect(cancelButton.length).toBe(1)
    expect(AppComponent.instance().state.isModalOpen).toBe(false)
})

test('User can put numeric value (invested amount) in the input', ()=> {
    const show = true
    const wrapper = shallow(<InvestForm show={show}/>)
    const investInputField = wrapper.find(FormControl)
    investInputField.simulate('change', {target: {
       name: 'investment',
       value: 50 
    }})
    
    expect(typeof wrapper.instance().state.investment).toBe("number")
    expect(wrapper.instance().state.investment).toBe(50)

})

test('User can click button labelled “Invest” which closes pop up', ()=> {
    const AppComponent = mount(<App/>)
    const spy = jest.spyOn(AppComponent.instance(), 'processInvestment');
    AppComponent.instance().setState({isModalOpen: true})

    const loan = loans[0]
    const show = true
    const wrapper = mount(<InvestForm selectedLoan={loan} processInvestment={spy} show={show}/>)
    const investButton = wrapper.find('#investButton').first()

    investButton.simulate('click')
    expect(investButton.length).toBe(1)
    expect(AppComponent.instance().state.isModalOpen).toBe(false)
})

test('When the "Invest" button is clicked, the available amount, for the loan User invested into, should decrease', ()=> {
    const AppComponent = mount(<App/>)
    const spy = jest.spyOn(AppComponent.instance(), 'processInvestment');
    AppComponent.instance().setState({isModalOpen: true})
    const currentAvailable = AppComponent.instance().state.loans[0].available

    const show = true
    const loan = loans[0]
    const wrapper = mount(<InvestForm selectedLoan={loan} processInvestment={spy} show={show}/>)

    const investInputField = wrapper.find(FormControl)
    const investButton = wrapper.find('#investButton').first()
    investInputField.simulate('change', {target: {
       name: 'investment',
       value: 50 
    }})

    investButton.simulate('click')

    const selectedLoan = wrapper.props().selectedLoan
    expect(wrapper.instance().state.investment).toBe(50)
    expect(parseFloat(selectedLoan.available.replace(/,/g, ''))).toBeLessThan(parseFloat(currentAvailable.replace(/,/g, '')))
    
})

test('When the "Invest" button is clicked, the total available number of investments should also adjust accordingly', ()=> {
    const AppComponent = shallow(<App/>)
    const spy = jest.spyOn(AppComponent.instance(), 'processInvestment');
    AppComponent.instance().setState({isModalOpen: true})
    const possibleInvestments = AppComponent.instance().state.possibleInvestments.length

    const show = true
    const loan = loans[0]
    const wrapper = mount(<InvestForm selectedLoan={loan} processInvestment={spy} show={show}/>)

    const investInputField = wrapper.find(FormControl)
    const investButton = wrapper.find('#investButton').first()
    investInputField.simulate('change', {target: {
       name: 'investment',
       value: 50 
    }})

    investButton.simulate('click')

    const selectedLoan = wrapper.props().selectedLoan
    expect(wrapper.instance().state.investment).toBe(50)
    expect(AppComponent.instance().state.possibleInvestments.length).toBeLessThan(possibleInvestments)
})


test('Investform component renders a snapshot properly', ()=>{
    const spy = jest.fn()
    const closeModal = jest.fn()
    const show = true
    const loan = loans[0]
    const tree = renderer.create(
        <InvestForm
        processInvestment={spy} 
        closeModal={closeModal} 
        selectedLoan={loan}/>).toJSON();
    expect(tree).toMatchSnapshot(); 
})