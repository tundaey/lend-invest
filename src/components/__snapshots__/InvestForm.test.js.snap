// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investform component renders a snapshot properly 1`] = `null`;
